File: ghost-translator.js
function encodeGhost(humanSpeak) {
    humanSpeak = humanSpeak.trim();

    if(humanSpeak == "") {
        return "Nothing entered!";
    }

    let ghostSpeak = "";

    for(let i = 0; i < humanSpeak.length; i++) {
        let strBinary = (humanSpeak.charCodeAt(i) ^ 96).toString(2);

        for(let j = 0; j < strBinary.length; j++) {
            if(strBinary[j] == "1") {
                ghostSpeak += "O";
            } else {
                ghostSpeak += "o";
            }
        }

        if(i != humanSpeak.length) {
            ghostSpeak += " ";
        }
    }

    enableCopy();
    return ghostSpeak;
}

function decodeGhost(ghostSpeak) {
    ghostSpeak = ghostSpeak.trim();

    if(ghostSpeak == "") {
        return "Nothing entered!";
    }

    let humanSpeak = "";

    let ghostWords = ghostSpeak.split(" ");
    for(let i = 0; i < ghostWords.length; i++) {
        let strBinary = "";

        for(let j = 0; j < ghostWords[i].length; j++) {
            if(ghostWords[i][j] == "O") {
                strBinary += "1";
            } else if(ghostWords[i][j] == "o") {
                strBinary += "0";
            } else {
                return "This ghost must have a speech impediment :(\n\nGhost speech consists only of O, o, and spaces.";
            }
        }

        humanSpeak += String.fromCharCode(parseInt(strBinary, 2) ^ 96);
    }

    enableCopy();
    return humanSpeak;
}

function enableCopy() {
    document.getElementById("copyButton").style.display = "inline";
}

async function copyText() {
    let tempInput = document.createElement("input");
    tempInput.value = document.getElementById("outputArea").innerText;
    tempInput.select();
    tempInput.setSelectionRange(0,99999);
    
    try {
        await navigator.clipboard.writeText(tempInput.value);
        document.getElementById("copyButton").value = "Copied! ✓";
        
        window.setTimeout(function() {
            document.getElementById("copyButton").value = "Copy results";
        }, 3000);
    } catch (err) {
        alert("Failed to copy to clipboard: ", err);
    }
}

File: ghost-translator.py
#!/usr/bin/env python3
from sys import argv

def main():
    if len(argv) > 2:
        if argv[1] == "-e" or argv[1] == "--encode":
            encodeGhost(argv[2])
        elif argv[1] == "-d" or argv[1] == "--decode":
            decodeGhost(argv[2])
        else:
            printUsage()
    else:
        printUsage()

def printUsage():
    print("Usage:")
    print("\tghost-translator -d (or --decode) \"OoOooo OoO OOooO OoooooO\"'  : Decode ghost-speak")
    print("\tghost-translator -e (or --encode) \"Hello!\"                     : Encode ghost-speak")
    print("\nExample:\n\tghost-translator -e \"Waddup!\"")

def encodeGhost(humanSpeak):
    ghostSpeak = ""

    for i in range(len(humanSpeak)):
        strBinary = bin(ord(humanSpeak[i])^96)[2:]

        for j in range(len(strBinary)):
            if strBinary[j] == "1":
                ghostSpeak += "O"
            else:
                ghostSpeak += "o"
        
        if i != len(humanSpeak):
            ghostSpeak += " "

    print(ghostSpeak)

def decodeGhost(strGhost):
    humanSpeak = ""

    ghostWords = strGhost.split()
    for i in range(len(ghostWords)):
        strBinary = ""
       
        for j in range(len(ghostWords[i])):
            if ghostWords[i][j] == "O":
                strBinary += "1"
            elif ghostWords[i][j] == "o":
                strBinary += "0"
            else:
                print(f"This ghost must have a speech impediment :(\n(Unexpected character in input: \"{ghostWords[i][j]}\")")
                exit()
            
        humanSpeak += str(chr(int(strBinary, 2)^96))

    print(humanSpeak)

main()
